### ISSUE TITLE
parser confused by autoloaded subroutine

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3486]
Migrated from [rt.perl.org#5918](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5918) (status was 'resolved')

Searchable as RT5918$

**Comment by p5pRT** [comment]
### From @gsar



Here's a much reduced test case of a curious problem\.

The bug is sensitive to the number of characters between the two
"EOT"s and the Getopt&#8203;::Long&#8203;::Configure\(\) call \(hence all those
hash marks\)\.  Assuming you preserved the whitespace in the test
case\, you should see the following when you run it&#8203;:

&nbsp;  String found where operator expected at \(eval 1\) line 20\, near "GetOptions\("a""
&nbsp;  xxx

This can be reproduced under Linux and Windows\.

\-\-\-\-8\<\-\-\-\-
use Getopt&#8203;::Long;
eval q\[
print \<\<EOT;
xxx
EOT

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

if \($x\) \{
 print \<\<EOT;
yyy
EOT
\}

\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\#\#\#\#\#\#\#\#\#\#\#\#\#\#
Getopt&#8203;::Long&#8203;::Configure\("ignore\_case"\);
GetOptions\("a"\, \\$a\);\];
print $@&#8203;;
\-\-\-\-8\<\-\-\-\-

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by build at Wed Aug 30 16:29:46 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.15, archname=i686-linux-thread-multi
    uname='linux churn.activestate.com 2.2.15 #1 smp thu may 25 03:22:16 pdt 2000 i686 unknown '
    config_args='-des -Dcc=gcc -Dcf_email=ActivePerl@ActiveState.com -Uinstallusrbinperl -Dusethreads -Duseithreads -Dprefix=/usr/local/ActivePerl-5.6'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-D_REENTRANT -fno-strict-aliasing'
    ccflags ='-D_REENTRANT -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lpthread -lc -lposix -lcrypt
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.0:
    /usr/local/ActivePerl-5.6/lib/5.6.0/i686-linux-thread-multi
    /usr/local/ActivePerl-5.6/lib/5.6.0
    /usr/local/ActivePerl-5.6/lib/site_perl/5.6.0/i686-linux-thread-multi
    /usr/local/ActivePerl-5.6/lib/site_perl/5.6.0
    /usr/local/ActivePerl-5.6/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/gsar
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/gsar/bin:/usr/local/bin:/usr/X11R6/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/X11R6/bin:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Mon\, 26 Feb 2001 21&#8203;:59&#8203;:03 PST\, gsar@&#8203;ActiveState\.com wrote&#8203;:
>The bug is sensitive to the number of characters between the two
>"EOT"s and the Getopt&#8203;::Long&#8203;::Configure\(\) call \(hence all those
>hash marks\)\.  Assuming you preserved the whitespace in the test
>case\, you should see the following when you run it&#8203;:
>
>  String found where operator expected at \(eval 1\) line 20\, near "GetOptions\("
>a""
>  xxx
>
>This can be reproduced under Linux and Windows\.

Here's the fix for this problem\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 8974 by gsar@moonru on 2001/03/01 16:28:21

	fix for bugid 20010226.008
	
	the problem was that some of the pointers (PL_last_lop and
	PL_last_uni specifically) into the lex buffers weren't correctly
	being invalidated when the buffer changed; this would leave the
	pointers pointing at an arbitrary location in the buffer if
	the buffer didn't need to be reallocated, or point into freed
	memory if the buffer had to be realloced
	
	TODO item for bugdb maintainers: check other seemingly random
	parser-related bugs--they might be cured by this

Affected files ...

... //depot/maint-5.6/perl/toke.c#47 edit

Differences ...

==== //depot/maint-5.6/perl/toke.c#47 (text) ====
Index: perl/toke.c
--- perl/toke.c.~1~	Thu Mar  1 08:26:42 2001
+++ perl/toke.c	Thu Mar  1 08:26:42 2001
@@ -359,6 +359,8 @@
     SAVEPPTR(PL_bufend);
     SAVEPPTR(PL_oldbufptr);
     SAVEPPTR(PL_oldoldbufptr);
+    SAVEPPTR(PL_last_lop);
+    SAVEPPTR(PL_last_uni);
     SAVEPPTR(PL_linestart);
     SAVESPTR(PL_linestr);
     SAVEPPTR(PL_lex_brackstack);
@@ -401,6 +403,7 @@
     SvTEMP_off(PL_linestr);
     PL_oldoldbufptr = PL_oldbufptr = PL_bufptr = PL_linestart = SvPVX(PL_linestr);
     PL_bufend = PL_bufptr + SvCUR(PL_linestr);
+    PL_last_lop = PL_last_uni = Nullch;
     SvREFCNT_dec(PL_rs);
     PL_rs = newSVpvn("\n", 1);
     PL_rsfp = 0;
@@ -546,6 +549,7 @@
 	    PL_oldoldbufptr = PL_oldbufptr = PL_bufptr = s = PL_linestart
 		= SvPVX(PL_linestr);
 	    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	    PL_last_lop = PL_last_uni = Nullch;
 
 	    /* Close the filehandle.  Could be from -P preprocessor,
 	     * STDIN, or a regular file.  If we were reading code from
@@ -1002,6 +1006,8 @@
     SAVEPPTR(PL_bufptr);
     SAVEPPTR(PL_oldbufptr);
     SAVEPPTR(PL_oldoldbufptr);
+    SAVEPPTR(PL_last_lop);
+    SAVEPPTR(PL_last_uni);
     SAVEPPTR(PL_linestart);
     SAVESPTR(PL_linestr);
     SAVEPPTR(PL_lex_brackstack);
@@ -1013,6 +1019,7 @@
     PL_bufend = PL_bufptr = PL_oldbufptr = PL_oldoldbufptr = PL_linestart
 	= SvPVX(PL_linestr);
     PL_bufend += SvCUR(PL_linestr);
+    PL_last_lop = PL_last_uni = Nullch;
     SAVEFREESV(PL_linestr);
 
     PL_lex_dojoin = FALSE;
@@ -1064,6 +1071,7 @@
 	PL_lex_inpat = 0;
 	PL_bufend = PL_bufptr = PL_oldbufptr = PL_oldoldbufptr = PL_linestart = SvPVX(PL_linestr);
 	PL_bufend += SvCUR(PL_linestr);
+	PL_last_lop = PL_last_uni = Nullch;
 	SAVEFREESV(PL_linestr);
 	PL_lex_dojoin = FALSE;
 	PL_lex_brackets = 0;
@@ -2519,6 +2527,7 @@
 	    sv_catpv(PL_linestr, "\n");
 	    PL_oldoldbufptr = PL_oldbufptr = s = PL_linestart = SvPVX(PL_linestr);
 	    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	    PL_last_lop = PL_last_uni = Nullch;
 	    if (PERLDB_LINE && PL_curstash != PL_debstash) {
 		SV *sv = NEWSV(85,0);
 
@@ -2547,10 +2556,12 @@
 		    sv_catpv(PL_linestr,";}");
 		    PL_oldoldbufptr = PL_oldbufptr = s = PL_linestart = SvPVX(PL_linestr);
 		    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+		    PL_last_lop = PL_last_uni = Nullch;
 		    PL_minus_n = PL_minus_p = 0;
 		    goto retry;
 		}
 		PL_oldoldbufptr = PL_oldbufptr = s = PL_linestart = SvPVX(PL_linestr);
+		PL_last_lop = PL_last_uni = Nullch;
 		sv_setpv(PL_linestr,"");
 		TOKEN(';');	/* not infinite loop because rsfp is NULL now */
 	    }
@@ -2593,6 +2604,7 @@
 		    sv_setpv(PL_linestr, "");
 		    PL_oldoldbufptr = PL_oldbufptr = s = PL_linestart = SvPVX(PL_linestr);
 		    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+		    PL_last_lop = PL_last_uni = Nullch;
 		    PL_doextract = FALSE;
 		}
 	    }
@@ -2607,6 +2619,7 @@
 	    av_store(CopFILEAV(PL_curcop),(I32)CopLINE(PL_curcop),sv);
 	}
 	PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	PL_last_lop = PL_last_uni = Nullch;
 	if (CopLINE(PL_curcop) == 1) {
 	    while (s < PL_bufend && isSPACE(*s))
 		s++;
@@ -2750,6 +2763,7 @@
 			    sv_setpv(PL_linestr, "");
 			    PL_oldoldbufptr = PL_oldbufptr = s = PL_linestart = SvPVX(PL_linestr);
 			    PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+			    PL_last_lop = PL_last_uni = Nullch;
 			    PL_preambled = FALSE;
 			    if (PERLDB_LINE)
 				(void)gv_fetchfile(PL_origfilename);
@@ -3632,7 +3646,7 @@
     case '\'':
 	s = scan_str(s,FALSE,FALSE);
         DEBUG_T( { PerlIO_printf(Perl_debug_log,
-                    "### Saw string in '%s'\n", s);
+                    "### Saw string before '%s'\n", s);
         } )
 	if (PL_expect == XOPERATOR) {
 	    if (PL_lex_formbrack && PL_lex_brackets == PL_lex_formbrack) {
@@ -3651,7 +3665,7 @@
     case '"':
 	s = scan_str(s,FALSE,FALSE);
         DEBUG_T( { PerlIO_printf(Perl_debug_log,
-                    "### Saw string in '%s'\n", s);
+                    "### Saw string before '%s'\n", s);
         } )
 	if (PL_expect == XOPERATOR) {
 	    if (PL_lex_formbrack && PL_lex_brackets == PL_lex_formbrack) {
@@ -3676,7 +3690,7 @@
     case '`':
 	s = scan_str(s,FALSE,FALSE);
         DEBUG_T( { PerlIO_printf(Perl_debug_log,
-                    "### Saw backtick string in '%s'\n", s);
+                    "### Saw backtick string before '%s'\n", s);
         } )
 	if (PL_expect == XOPERATOR)
 	    no_op("Backticks",s);
@@ -6435,6 +6449,7 @@
 	sv_setsv(PL_linestr,herewas);
 	PL_oldoldbufptr = PL_oldbufptr = PL_bufptr = s = PL_linestart = SvPVX(PL_linestr);
 	PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	PL_last_lop = PL_last_uni = Nullch;
     }
     else
 	sv_setpvn(tmpstr,"",0);   /* avoid "uninitialized" warning */
@@ -6446,6 +6461,7 @@
 	}
 	CopLINE_inc(PL_curcop);
 	PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	PL_last_lop = PL_last_uni = Nullch;
 #ifndef PERL_STRICT_CR
 	if (PL_bufend - PL_linestart >= 2) {
 	    if ((PL_bufend[-2] == '\r' && PL_bufend[-1] == '\n') ||
@@ -6804,6 +6820,7 @@
 
 	/* having changed the buffer, we must update PL_bufend */
 	PL_bufend = SvPVX(PL_linestr) + SvCUR(PL_linestr);
+	PL_last_lop = PL_last_uni = Nullch;
     }
 
     /* at this point, we have successfully read the delimited string */
@@ -7320,6 +7337,7 @@
 	    s = filter_gets(PL_linestr, PL_rsfp, 0);
 	    PL_oldoldbufptr = PL_oldbufptr = PL_bufptr = PL_linestart = SvPVX(PL_linestr);
 	    PL_bufend = PL_bufptr + SvCUR(PL_linestr);
+	    PL_last_lop = PL_last_uni = Nullch;
 	    if (!s) {
 		s = PL_bufptr;
 		yyerror("Format not terminated");
End of Patch.
```
</details>
